r-package/grf/R/average_late.R

```diff
@@ -0,0 +1,151 @@
+#'
+#' @importFrom stats coef lm predict var weighted.mean
+#' @export
+average_late <- function(forest,
+                         compliance.score = NULL,
+                         subset = NULL) {
+  cluster.se <- length(forest$clusters) > 0
+
+  if (!("instrumental_forest" %in% class(forest))) {
+      stop(paste(
+          "Average conditional local average treatment effect estimation",
+          "only implemented for instrumental_forest."
+      ))
+  }
+  
+  if (!all(forest$Z.orig %in% c(0, 1))) {
+      stop(paste(
+          "Average conditional local average treatment effect estimation",
+          "only implemented for binary instruments."
+      ))
+  }
+
+  if (is.null(subset)) {
+    subset <- 1:length(forest$Y.hat)
+  }
+
+  if (class(subset) == "logical" & length(subset) == length(forest$Y.hat)) {
+    subset <- which(subset)
+  }
+
+  if (!all(subset %in% 1:length(forest$Y.hat))) {
+    stop(paste(
+      "If specified, subset must be a vector contained in 1:n,",
+      "or a boolean vector of length n."
+    ))
+  }
+
+  clusters <- if (cluster.se) {
+    forest$clusters
+  } else {
+    1:length(forest$Y.orig)
+  }
+  observation.weight <- observation_weights(forest)
+  
+  # The compliance forest estimates the effect of the "treatment" Z
+  # on the "outcome" W.
+  if (is.null(compliance.score)) {
+      compliance.forest <- causal_forest(forest$X.orig,
+                                         Y=forest$W.orig,
+                                         W=forest$Z.orig,
+                                         Y.hat=forest$W.hat,
+                                         W.hat=forest$Z.hat,
+                                         sample.weights = forest$sample.weights,
+                                         clusters = clusters)
+      compliance.score <- predict(compliance.forest)$predictions
+  }
+
+  # Only use data selected via subsetting.
+  subset.W.orig <- forest$W.orig[subset]
+  subset.W.hat <- forest$W.hat[subset]
+  subset.Z.orig <- forest$Z.orig[subset]
+  subset.Z.hat <- forest$Z.hat[subset]
+  subset.Y.orig <- forest$Y.orig[subset]
+  subset.Y.hat <- forest$Y.hat[subset]
+  subset.compliance.score <- compliance.score[subset]
+  subset.tau.hat.pointwise <- predict(forest)$predictions[subset]
+  subset.clusters <- clusters[subset]
+  subset.weights <- observation.weight[subset]
+
+  if (min(subset.Z.hat) <= 0.01 || max(subset.Z.hat) >= 0.99) {
+    rng <- range(subset.W.hat)
+    warning(paste0(
+      "Estimated treatment propensities take values between ",
+      round(rng[1], 3), " and ", round(rng[2], 3),
+      " and in particular get very close to 0 or 1. ",
+      "Poor overlap may hurt perfmance for average conditional local average",
+      "treatment effect estimation."
+    ))
+  }
+  
+  if (min(subset.compliance.score) <= 0.01 * sd(subset.W.orig)) {
+      warning(paste0(
+          "The instrument appears to be weak, with some compliance scores as",
```

Comment:
Space after the word "as", here and on the next line. The warnings currently come out as
```
The instrument appears to be weak, with some compliance scores aslow as0
```